File: php/src/Permutation.php
<?php

function lox617_catmap(
    array &$block
): void {

    $temp =
        array_fill(
            0,
            8,
            0
        );

    for($i = 0; $i < 8; $i++) {

        for($j = 0; $j < 8; $j++) {

            $bit =
                ($block[$i] >> $j) & 1;

            $newX =
                (2 * $i + $j) % 8;

            $newY =
                ($i + $j) % 8;

            $temp[$newX] |=
                ($bit << $newY);
        }
    }

    for($i = 0; $i < 8; $i++) {

        $block[$i] =
            $temp[$i] & 0xFF;
    }
}

File: rust/src/permutation.rs
pub fn lox617_catmap(
    block: &mut [u8; 8]
) {

    let mut temp =
        [0u8; 8];

    for i in 0..8 {

        for j in 0..8 {

            let bit =
                (block[i] >> j)
                & 1;

            let new_x =
                (2 * i + j) % 8;

            let new_y =
                (i + j) % 8;

            temp[new_x] |=
                bit << new_y;
        }
    }

    *block = temp;
}